File: frontend/lib/api/review.ts
import { api } from "./client";

export interface CriterionReview {
  criterion_id:     string;
  name:             string;
  description:      string;
  max_marks:        number;
  display_order:    number;
  ai_score:         number | null;
  ai_justification: string | null;
  override_score:   number | null;
  override_note:    string | null;
  effective_score:  number;
}

export interface FinalisedGradeOut {
  id:              string;
  submission_id:   string;
  teacher_id:      string;
  total_score:     number;
  teacher_comment: string | null;
  is_published:    boolean;
  finalised_at:    string;
}

export interface SubmissionReview {
  id:              string;
  assignment_id:   string;
  student_id:      string;
  student_name:    string | null;
  submission_type: string;
  status:          string;
  text_content:    string | null;
  file_name:       string | null;
  submitted_at:    string;
  assignment_title: string;
  question_text:    string;
  max_marks:        number;
  total_ai_score:     number | null;
  formative_feedback: string | null;
  criteria:           CriterionReview[];
  effective_total:    number;
  finalised_grade:    FinalisedGradeOut | null;
}

export const reviewApi = {
  get: (submissionId: string): Promise<SubmissionReview> =>
    api.get(`/submissions/${submissionId}/review`),

  setOverride: (
    submissionId:      string,
    rubricCriterionId: string,
    score:             number,
    note?:             string,
  ): Promise<SubmissionReview> =>
    api.post(`/submissions/${submissionId}/overrides`, {
      rubric_criterion_id: rubricCriterionId,
      overridden_score:    score,
      override_note:       note ?? null,
    }),

  removeOverride: (submissionId: string, criterionId: string): Promise<SubmissionReview> =>
    api.delete(`/submissions/${submissionId}/overrides/${criterionId}`),

  finalise: (submissionId: string, teacherComment?: string): Promise<SubmissionReview> =>
    api.post(`/submissions/${submissionId}/finalise`, { teacher_comment: teacherComment ?? null }),

  publishGrade: (submissionId: string): Promise<SubmissionReview> =>
    api.post(`/submissions/${submissionId}/publish-grade`),
};

File: backend/app/schemas/review.py
import uuid
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


class OverrideIn(BaseModel):
    rubric_criterion_id: uuid.UUID
    overridden_score:    float = Field(..., ge=0)
    override_note:       str | None = None


class FinaliseRequest(BaseModel):
    teacher_comment: str | None = None


class CriterionReviewOut(BaseModel):
    criterion_id:     uuid.UUID
    name:             str
    description:      str
    max_marks:        int
    display_order:    int
    ai_score:         float | None
    ai_justification: str | None
    override_score:   float | None
    override_note:    str | None
    effective_score:  float


class FinalisedGradeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id:              uuid.UUID
    submission_id:   uuid.UUID
    teacher_id:      uuid.UUID
    total_score:     float
    teacher_comment: str | None
    is_published:    bool
    finalised_at:    datetime


class SubmissionReviewOut(BaseModel):
    id:              uuid.UUID
    assignment_id:   uuid.UUID
    student_id:      uuid.UUID
    student_name:    str | None
    submission_type: str
    status:          str
    text_content:    str | None
    file_name:       str | None
    submitted_at:    datetime
    # Assignment info
    assignment_title: str
    question_text:    str
    max_marks:        int
    # AI grading
    total_ai_score:     float | None
    formative_feedback: str | None
    # Per-criterion review
    criteria:        list[CriterionReviewOut]
    effective_total: float
    # Finalised grade (present once teacher finalises)
    finalised_grade: FinalisedGradeOut | None = None
